### PR TITLE
Filament autoload

### PR DESCRIPTION
**Pull request by Hagridpon3**

### CODE REVIEW

File: src/common/thread_measurement.cpp [hunk 1]
```diff
@@ -21,20 +20,48 @@ void StartMeasurementTask(void const *argument) {
...
+        if (next_fs_cycle <= now) {
             fs_cycle();
             // call fs_cycle every ~50 ms
             next_fs_cycle = now + 50;
         }
 
+        // sample filament autoload
+        // call every ~100ms
+        marlin_update_vars(MARLIN_VAR_MSK_FS);
+        marlin_vars_t *vars = marlin_update_vars(MARLIN_VAR_SD_PRINT);
```
Comment (dragomirecky) [inline src/common/thread_measurement.cpp:42]: /question/ I am not really familiar with marlin_client's api, but shouldn't it be possible to read this in one go?

```C
marlin_vars_t *vars = marlin_update_vars(MARLIN_VAR_SD_PRINT | MARLIN_VAR_MSK_FS);
```

File: src/common/thread_measurement.cpp [hunk 2]
```diff
@@ -53,6 +80,6 @@ void StartMeasurementTask(void const *argument) {
             next_sg_cycle = now + next_delay;
         }
 
-        osDelay(checkTimestampsAscendingOrder(next_fs_cycle, next_sg_cycle) ? next_fs_cycle - now : next_sg_cycle - now);
+        osDelay(std::max({ next_fs_cycle, next_sg_cycle, next_fs_autoload_cycle }) - now);
```
Comment (dragomirecky) [inline src/common/thread_measurement.cpp:83]: `checkTimestampsAscendingOrder` properly handled the case of uint32_t overflowing; this ignores the problem :)

Btw – shouldn't it be std::min? You want to wake up with the closest event, not on the one being the furthest in the future.

File: src/common/thread_measurement.cpp (hunk 1 above)
Comment (dragomirecky) [reply src/common/thread_measurement.cpp:42]: /question/ Why don't we check for the `next_fs_autoload_cycle` first and then check whether autoload is enabled? Updating marlin vars seems like more expensive operation we might not want to do on every cycle here.

File: src/common/thread_measurement.cpp [hunk 3]
```diff
@@ -5,13 +5,12 @@
...
+#include "marlin_server.h"
 #include "trinamic.h"
 #include "stm32f4xx_hal.h"
+#include "eeprom.h"
 
-static inline bool checkTimestampsAscendingOrder(uint32_t a, uint32_t b) {
-    uint32_t u = (b - a);
-    return !(u & 0x80000000u);
-}
+static bool current_detect_filament_insert = false;
```
Comment (dragomirecky) [inline src/common/thread_measurement.cpp:13]: /question/ Why isn't this a local variable? I don't see it being used anywhere else then in StartMeasurementTask().

File: src/common/thread_measurement.cpp (hunk 1 above)
Comment (Hagridpon3) [reply src/common/thread_measurement.cpp]: Solved.

File: src/common/thread_measurement.cpp (hunk 3 above)
Comment (Hagridpon3) [reply src/common/thread_measurement.cpp]: Solved

File: src/gui/screen_menu_settings.cpp [hunk 4]
```diff
@@ -69,7 +69,12 @@ bool MI_FILAMENT_SENSOR::consumeNotConnected() {
 }
 
 void MI_FILAMENT_SENSOR::OnChange(size_t old_index) {
-    old_index == 1 ? fs_disable() : fs_enable();
```
Comment (RadekVana) [inline src/gui/screen_menu_settings.cpp:72]: unnecessary change, pls remove

File: src/common/thread_measurement.cpp [hunk 5]
```diff
@@ -33,6 +40,29 @@ void StartMeasurementTask(void const *argument) {
             next_fs_cycle = now + 50;
         }
 
+        // sample filament autoload
```
Comment (RadekVana) [inline src/common/thread_measurement.cpp:43]: move all changed code from this file into filament_sensor.cpp
create static function `autoload_loop` and call it inside `fs_cycle();` faster cycle 50ms for autoload is irrelevant

**Review by RadekVana — CHANGES_REQUESTED** [review]

File: src/gui/screen_menu_settings.cpp (hunk 4 above)
Comment (Hagridpon3) [reply src/gui/screen_menu_settings.cpp:72]: Solved

File: src/common/thread_measurement.cpp (hunk 5 above)
Comment (Hagridpon3) [reply src/common/thread_measurement.cpp]: Solved

**Review by RadekVana — APPROVED** [review]
